tools/lcdtools/paneld/paneld.8
.\" $Id: paneld.8 181 2005-05-07 06:51:15Z pdh $
.\"
.\" Copyright (C) 2004, 2005  Martin Michlmayr <tbm@cyrius.com>
.\" This manual is freely distributable under the terms of the GPL.
.\" It was originally written for Debian GNU/Linux (but may be used
.\" by others).
.\"

.TH PANELD 8 "July 2004"

.SH NAME
paneld \- Cobalt admin menu for the LCD panel

.SH SYNOPSIS
.B paneld
[\-d] [\-h] [\-\-] [\fISTRINGs\fR]

.SH DESCRIPTION
.PP
.B paneld
is an admin tool for the LCD panel of Cobalt machines.  By default, it will
display the current time and optionally a message.  When you hold the
.B enter
or
.B select
button for a couple of seconds you will get an admin menu.  The menu will
allow you to either halt or reboot your Cobalt machine.

It is possible to pass a set of strings to paneld.  In this case,
.B paneld
will loop through each argument and display the string for a few seconds
in the second line below the time.  It will proceed with the next argument
and loop around when the last argument is reached.

If called with the
.B \-d
option
.B paneld
will fork itself into the background immediately.

The
.B \-h
option will make
.B paneld
use scrolling horizontal menus rather than the usual vertical menu.

.SH AUTHOR
.PP
Peter Horton <pdh@colonel\-panic.org>



CHANGES
# $Id: CHANGES 195 2006-08-10 22:03:35Z pdh $

CoLo 1.22 (2006-08-10)
----------------------

*	Fix build failure on GCC 4.1.

*	Tweaked INSTALL.
	[Martin Michlmayr]

*	Fixed 'netcon' to not cause an assert in UDP code.

*	Added support for CF cards in CF/IDE adapters (not tested).

CoLo 1.21 (2006-01-22)
----------------------

*	Merged "screen" and "minicom" keymaps.

*	Added "delete word" to line editor ('<ctrl>w').

*	Added 'relocate' command which relocates the initrd image to follow the
	kernel image. Required to use initrd images with current 2.6.x kernels.

*	'mount' command can now be aborted.

CoLo 1.20 (2006-01-16)
----------------------

*	Fix menu cancel/abort/error returns.

*	Update NV format to include keymap.

*	Added keymap for "screen".

*	Wait for network interface transmit ring to drain when disabling network
	interface. This should improve net console output.

*	Added header file dependency checking to stage1, stage2 and chain.

CoLo 1.19 (2005-04-03)
----------------------

*	Fixed stupid bug in ordering of LCD initialisation. This would cause
	installs of 1.17 or 1.18 which weren't an upgrade from 1.16 to crash
	early in stage2.

*	Added "README.tools" to tools directory.

*	Renamed 'md5rom' to 'copy-rom' and added the ability to save a copy of the
	ROM to a file.

*	Moved PCI timeout / retry counter initialisation to PCI initialisation
	code. We now use a retry count of 255 rather than retrying forever. The
	kernel needs a similiar fix.

*	Improved PCI bus scan.

CoLo 1.18 (2005-03-22)
----------------------

*	Added UDP network console support. This adds a 'netcon' command and
	associated variables, and a new boot menu item "Network Shell".

*	Added "ncon" to tools directory. Very simple net console client.

*	PCI serial support switch moved to NV flag 7.

CoLo 1.17 (2005-02-20)
----------------------

*	DHCP client now broadcasts DHCPREQUEST packets. It also includes the
	"server identifier" option from the server in the request. Fixed bug where
	we weren't appending an "end options" tag to the options. [RFC2131]

*	Dropped "no matching socket" diagnostic message.

*	Simplified display of memory size on boot.

*	Added "Cancel" option to 'paneld' menus.

*	Added horizontal scrolling menus to 'paneld' (Controlled by the '-h'
	option).

*	Added horizontal scrolling menus to shell. Controlled by an NV flag (see
	'nvflags' command in README.shell).

*	Reset menu timeout when a button is pressed.

*	Removed ARP diagnostic messages.

CoLo 1.16 (2004-12-16)
----------------------

*	Force menu choice of console on/off on upgrade from earlier versions.

*	Fix 'flash-tool' and 'flash' command to work with TMS29F040 Flash.

CoLo 1.15 (2004-11-28)
----------------------

*	Added link detection to network driver.

*	Added support for 21041 network controller (2700WG units).

*	Added clock speed detection (stage2 only) to ensure timeouts and delays are
	consistent regardless of CPU speed.

*	Added new command 'sleep'.

CoLo 1.14 (2004-10-26)
----------------------

*	Cleaned up build system and added cross compile support.
	[cross compile suggestion / gcc-3.4 link fix - Sebastien Vajda]
	[compiler flag suggestions - Thiemo Seufer]

*	Added new example boot script 'menu.colo' showing new features.

*	Variable expansion in the shell is now carried out before argument
	splitting.  This means variables can now span arguments e.g.

		var message "'first' 'second'"
		lcd {message}

*	Reworked the LCD menu command. The 'menu' command has been removed and
	replaced with a new command, 'select'. The new command justs sets a
	variable to the index of the selected item which can then be used with
	'goto'.

*	Added 'noop' command, which does nothing. When combined with the 'onfail'
	command it can be used for checking the existence of variables.

*	Added script commands 'goto', 'onfail', 'exit' and 'abort'.

*	Added positional marks in scripts using the '@' character.

*	Scripts can now be nested.

*	The 'console-speed' variable is now only available when the serial port is
	enabled.

*	Fixed up support for booting 64-bit kernels (32-bit support was reorganised
	too). 32-bit kernels must be linked for KSEG0 or KSEG1, 64-bit kernels for
	CKSEG0, CKSEG1 or XKPHYS.

*	Ensure we read from Galileo PCI RETRY/TIMEOUT register before writing to
	it. A comment in the kernel source indicates this might be necessary to
	stop Galileo locking up on early units.

*	Added a '-n' command line option to 'putlcd' which allows the user to write
	to a single line of the LCD without affecting the other line.

*	Added big fat warning to 'flash-tool' (just in case).

CoLo 1.13 (2004-08-30)
----------------------

*	Speeded up raw LCD access in LCD tools library.

*	Added 'on' and 'off' options to the 'serial' command.

*	Added 'colo-perm' and 'md5rom' to build.

*	Added missing include to paneld.c.

CoLo 1.12 (2004-08-20)
----------------------

*	The LCD tools 'putlcd', 'paneld' and 'e2fsck-lcd' are now built around a
	common library. The library has support for using direct access to the LCD
	via /dev/mem (the default) or driver access via the Cobalt LCD driver.

	To make the library use the Cobalt LCD driver you need to export an
	environment variable LCD_DEVICE with the name of the device. If LCD_DEVICE
	is an empty string the library will use the default device name, /dev/lcd.

*	Added a '-d' command line option to 'paneld' which makes it fork itself
	into the background on startup.

*	Added 'menu' boot shell command for implementing LCD menus in scripts.
	This feature makes it possible to choose from a list of multiple
	kernels/configurations to boot (see example in README.shell).

*	Fixed a couple of off by one errors in the 'boot' command.

*	Added 'console disable' to the non-volatile flags. This defers initialising
	the serial port until the boot shell is entered and should make it possible
	to use CoLo on units which don't have a serial port. This flag will be set

tools/lcdtools/paneld/paneld.c
/*
 * (C) P.Horton 2004,2005,2006
 *
 * $Id: paneld.c 186 2006-01-17 23:03:58Z pdh $
 *
 * This code is covered by the GNU General Public License. For details see the file "COPYING".
 */

#include <unistd.h>
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <fcntl.h>
#include <string.h>
#include <errno.h>
#include <time.h>
#include <sys/mman.h>
#include <sys/time.h>

#include "liblcd.h"

#define APP_NAME					"paneld"

#define LCD_MENU_ERROR			(-1)
#define LCD_MENU_TIMEOUT		(-2)
#define LCD_MENU_CANCEL			(-3)

#define BTN_DEBOUNCE				(50 * 1000)
#define UPDATE_TICK				(500 * 1000)
#define UPDATE_PERIOD_COUNT	(5 * 1000 * 1000 / UPDATE_TICK)
#define MENU_SELECT_COUNT		(1000 * 1000 * 3 / 2 / UPDATE_TICK)
#define MENU_TIMEOUT				(10 * 1000 * 1000)
#define MENU_MESSAGE				(5 * 1000 * 1000)

#define ELEMENTS(x)				(sizeof(x)/sizeof((x)[0]))

static const char *menu_options[] =
{
	"ADMIN MENU",
	"Halt",
	"Reboot",
	"Cancel",
};

static const char *menu_messages[][2] =
{
	{ "Shutting down", "  for halt.", },
	{ "Shutting down", " for reboot.", },
};

static const char *action_halt[] =
{
	"/sbin/shutdown", "-a", "-t1", "-h", "now", NULL,
};

static const char *action_reboot[] =
{
	"/sbin/shutdown", "-a", "-t1", "-r", "now", NULL,
};

static const char **menu_actions[] =
{
	action_halt,
	action_reboot,
};

const char *getapp(void)
{
	return APP_NAME;
}

static const char *lcd_symbols(const char *def)
{
	static const uint8_t arrow[][8] =
	{
		{ 0x01, 0x03, 0x07, 0x0f, 0x07, 0x03, 0x01, 0x00, },
		{ 0x10, 0x18, 0x1c, 0x1e, 0x1c, 0x18, 0x10, 0x00, },
	};
	static char buf[3];

	buf[0] = def[0];
	buf[1] = def[1];

	if(lcd_prog(1, arrow[0]))
		buf[0] = '\001';

	if(lcd_prog(2, arrow[1]))
		buf[1] = '\002';

	return buf;
}

static void lcd_centre(char *buf, const char *str, unsigned siz)
{
	unsigned ofs, len, idx;

	ofs = 0;
	len = strlen(str);
	if(len < siz)
		ofs = (siz - len) / 2;

	for(idx = 0; idx < siz; ++idx) {

		buf[idx] = ' ';
		if(idx >= ofs && idx < ofs + len)
			buf[idx] = *str++;
	}
}

static void lcd_scroll(const char *str, int dir)
{
	static char lcd[LCD_WIDTH - 2];
	char buf[LCD_WIDTH - 2];
	unsigned idx, num;
	const char *sym;

	lcd_centre(buf, str, sizeof(buf));

	if(!dir) {

		memcpy(lcd, buf, sizeof(lcd));

		sym = lcd_symbols("<>");

		lcd_puts(1, 0, 1, &sym[0]);
		lcd_puts(1, 1, sizeof(lcd), lcd);
		lcd_puts(1, 1 + sizeof(lcd), 1, &sym[1]);

		return;
	}

	for(num = sizeof(lcd);;) {

		if(dir < 0) {

			for(idx = 1; idx < sizeof(lcd); ++idx)
				lcd[idx - 1] = lcd[idx];
			lcd[sizeof(lcd) - 1] = buf[sizeof(lcd) - num];

		} else {

			for(idx = sizeof(lcd); --idx;)
				lcd[idx] = lcd[idx - 1];
			lcd[0] = buf[num - 1];
		}

		lcd_puts(1, 1, sizeof(lcd), lcd);

		if(!--num)
			break;

		usleep(30 * 1000);
	}
}

static int lcd_menu_horz(const char **options, unsigned count, unsigned timeout)
{
	unsigned exp, sel, btn, prv;
	struct timeval mark, now;
	char buf[LCD_WIDTH];
	int dir;

	if(count < 2)
		return LCD_MENU_ERROR;

	lcd_centre(buf, options[0], sizeof(buf));
	lcd_puts(0, 0, sizeof(buf), buf);

	lcd_scroll(options[1], 0);

	prv = btn_read();

	for(sel = 1;;) {

		gettimeofday(&mark, NULL);
		
		for(;;) {

			if(timeout) {

				gettimeofday(&now, NULL);

				exp = (now.tv_sec - mark.tv_sec) * 1000000 + now.tv_usec - mark.tv_usec;

				if(exp >= timeout)
					return LCD_MENU_TIMEOUT;
			}

			btn = btn_read();

			btn ^= prv;
			prv ^= btn;
			btn &= prv;

			if(btn & (BTN_ENTER | BTN_SELECT))
				return sel - 1;

			if(btn & (BTN_UP | BTN_DOWN))
				return LCD_MENU_CANCEL;

			if(btn & (BTN_LEFT | BTN_RIGHT)) {

				dir = 1;
				if(btn & BTN_RIGHT)
					dir = -1;

				sel -= dir;
				if(sel < 1 || sel >= count)
					sel = count - sel - dir;

				lcd_scroll(options[sel], dir);

			} else

				usleep(BTN_DEBOUNCE);
		}
	}

	return LCD_MENU_ERROR;
}

static int lcd_menu_vert(const char **options, unsigned count, unsigned timeout)
{
	unsigned exp, row, top, btn, prv;
	struct timeval mark, now;
	const char *sym;
	char buf[2];

	if(count < 2)
		return LCD_MENU_ERROR;

	sym = lcd_symbols("][");

	prv = btn_read();

	buf[1] = '\0';
	row = 1;
	top = 0;

	for(;;) {

		lcd_puts(0, 0, 1, row ? " " : &sym[1]);
		lcd_puts(0, 1, LCD_WIDTH - 2, options[top]);
		lcd_puts(0, LCD_WIDTH - 1, 1, row ? " " : &sym[0]);

		lcd_puts(1, 0, 1, row ? &sym[1] : " ");
		lcd_puts(1, 1, LCD_WIDTH - 2, options[top + 1]);
		lcd_puts(1, LCD_WIDTH - 1, 1, row ? &sym[0] : " ");

		gettimeofday(&mark, NULL);
		
		for(;;) {

			if(timeout) {

				gettimeofday(&now, NULL);

				exp = (now.tv_sec - mark.tv_sec) * 1000000 + now.tv_usec - mark.tv_usec;

				if(exp >= timeout)
					return LCD_MENU_TIMEOUT;
			}

			btn = btn_read();

			btn ^= prv;
			prv ^= btn;
			btn &= prv;

			if(btn & (BTN_RIGHT | BTN_ENTER | BTN_SELECT))
				return top + row - 1;

			if(btn & BTN_LEFT)
				return LCD_MENU_CANCEL;

			if(btn & BTN_UP) {
				if(top) {
					--top;
					row = 1;
					break;
				}
				if(row != 1) {
					row = 0;
					break;
				}
			}

			if(btn & BTN_DOWN) {
				if(top + 2 < count) {
					++top;
					row = 0;
					break;
				} 
				if(!row) {
					row = 1;
					break;
				}
			}

			usleep(BTN_DEBOUNCE);
		}
	}
}

static void usage(void)
{
	puts("usage: " APP_NAME " [ -d ] [ -h ]");
	exit(1);
}

int main(int argc, char *argv[])
{
	static int (* const menu[])(const char **, unsigned, unsigned) =
	{
		lcd_menu_vert,
		lcd_menu_horz,
	};
	static char text[32];

	unsigned count, update, active, roller;
	int opt, daemon, which, mtype;
	time_t wall;

	daemon = 0;
	mtype = 0;

	opterr = 0;

	while((opt = getopt(argc, argv, "dh")) != -1)
		switch(opt) {
			case 'd':
				daemon = 1;
				break;
			case 'h':
				mtype = 1;
				break;
			default:
				usage();
		}

	if(!lcd_open(NULL))
		return -1;

	if(daemon)

		switch(fork()) {

			case -1:
				fprintf(stderr, APP_NAME ": failed to fork (%s)\n", strerror(errno));
				return -1;

			case 0:
				setsid();
				break;

			default:
				return 0;
		}

	for(roller = optind;;) {

		active = MENU_SELECT_COUNT;
		update = 0;

		for(count = UPDATE_PERIOD_COUNT;; ++count) {

			if(count - update >= UPDATE_PERIOD_COUNT) {

				update += UPDATE_PERIOD_COUNT;

				time(&wall);
				strftime(text, sizeof(text), "%a %b %d %H:%M", localtime(&wall));
				lcd_puts(0, 0, LCD_WIDTH, text);

				if(roller < argc) {

					lcd_puts(1, 0, LCD_WIDTH, argv[roller++]);

					if(roller == argc)
						roller = optind;

				} else

					lcd_puts(1, 0, LCD_WIDTH, "");
			}

			if(btn_read() & (BTN_SELECT | BTN_ENTER)) {
				if(active < MENU_SELECT_COUNT && ++active == MENU_SELECT_COUNT)
					break;
			} else
				active = 0;

			usleep(UPDATE_TICK);
		}

		which = menu[mtype](menu_options, ELEMENTS(menu_options), MENU_TIMEOUT);

		if(which >= 0 && which < ELEMENTS(menu_actions)) {

			lcd_puts(0, 0, LCD_WIDTH, menu_messages[which][0]);
			lcd_puts(1, 0, LCD_WIDTH, menu_messages[which][1]);

			switch(fork()) {

				case -1:
					fprintf(stderr, APP_NAME ": failed to fork (%s)\n", strerror(errno));
					break;

				case 0:
					lcd_close();

					execv(menu_actions[which][0], (char **) menu_actions[which]);

					fprintf(stderr, APP_NAME ": failed to exec \"%s\" (%s)\n", menu_actions[which][0], strerror(errno));
					return -1;
			}

			usleep(MENU_MESSAGE);
		}
	}

	return 0;
}

/* vi:set ts=3 sw=3 cin: */
